k8s fixes

zeenix opened the pull request:
Same as https://github.com/kinvolk/habitat/pull/13 but against my unmerged parent branch for each review. :)

mgamini committed:
adding Getting Started with Node blog post

Signed-off-by: Garrett Amini <kadjar@gmail.com>

adding author

Signed-off-by: Garrett Amini <kadjar@gmail.com>

Commit:
pkg-export-k8s: Don't throw programming error to user

ManifestJson::to_string() only returns errors if the crate programmer
makes a mistake (e.g messed-up the manifest template or doesn't check the
user input) so instead of forwarding the error to caller, best just give
up with a message on the console.

This also means that we can just implement Into<String> trait for
ManifestJson so users can use well-known Rust API. Unfortunately
Into::into() takes the ownership of self so the using methods had to be
changed to take ownership of self as well.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Remove random string from service name

This is mainly a reversion of 7a275f106f558bdd8a5553d980d7ac77e6ae8a19
"k8s exporter: Append a random suffix to the name of Habitat object". The
reason this was added was to be able to run multiple versions of the
habitat service at the same time for making channel promotion possible.
However, this approach was later dropped and the corresponding changes to
(kubernetes) Habitat operator were never done.

Currently addition of random string only means that the initial
configuration file is put in the wrong location in the kubernetes pod.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add version & release to names

When available add version and release to the service name to make it
explicit which version and release the generated manifest is for and also
to allow running different versions and releases of the same service in the
same namespace.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

lilic reviewed (approved):
lgtm

zeenix commented:
Thanks @LiliC . As @krnowak pointed out in #13, we are assuming that docker exporter will generate a specific tag and now that I checked the code, it will not produce the assumed tag if user passes `--no-tag-version` on the CLI. Moreover, even `:latest` tag can be disabled with a CLI arg. :( So I think it's best if k8s exporter ensure docker exporter used the same tags.

Commit:
pkg-export-docker: Drop a redundant 'if' block

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

asymmetric commented on components/pkg-export-kubernetes/src/manifestjson.rs:
```diff
@@ -71,30 +68,29 @@ impl ManifestJson {
-    pub fn into_string(&self) -> Result<String> {
+    fn into(self) -> String {
+        // The Result::expect() usage in this function is justied by the fact that errors can only
+        // come from crate programmer (e.g messed-up the manifest template or don't check the user
```
Sorry to be a grammar freak again, but: I don't understand what the subject is in "e.g messed-up the manifest template or don't check the user input". So maybe reword it to something like "e.g. if the manifest template has syntax errors or user input is invalidat".

zeenix replied:
sure.

asymmetric commented on components/pkg-export-helm/src/main.rs:
```diff
@@ -54,7 +54,7 @@ fn main() {
 
 fn export_for_cli_matches(ui: &mut UI, matches: &clap::ArgMatches) -> Result<()> {
-    let mut chart = Chart::new_for_cli_matches(ui, matches)?;
+    let chart = Chart::new_for_cli_matches(ui, matches)?;
```
It's good we can make something immutable, but one thing I don't understand: why was it necessary for it to be mutable before? And what changed in between?

zeenix replied:
because of `chart.genarate()` not taking ownership and hence needing mutability IIRC. I only removed mut cause rustc informed me it's not needed anymore.

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -129,7 +129,7 @@ impl Manifest {
     /// Generates the manifest as a string and writes it to `write`.
     pub fn generate(&mut self, write: &mut Write) -> Result<()> {
-        let out = ManifestJson::new(&self).into_string()?;
+        let out: String = ManifestJson::new(&self).into();
```
A bit of a design smell with this IMO is that you end up having to use a type annotation anyway in the only place you use the `.into()`. I find `.into()` makes sense when you can avoid to do that cause the compiler manages to infer the type.

Another approach would be the [`ToString`](https://doc.rust-lang.org/alloc/string/trait.ToString.html#tymethod.to_string) trait, which you get by implementing the `Display` trait, which in turn could be handy as well.

My 2 cents.

zeenix replied:
I don't think you decide to use Into trait or not base weather not a paricular usage of it will have to use type annotation. if the usage of 'out' afterwards would have made the type obvious to the compiler, we wouldn't need to annotate.

I did indeed go with implementing Display train actually but it felt wrong since that's mostly to create a user-visible string representation to show to user in messages and stuff. Although at that time, I hadn't decided that this method shouldn't return a Result. Now that I think of it, ToString() might be better indeed cause it's best to be explicit if possible but the docs discouraging implementations of it, makes me a bit vary of this trait.

asymmetric replied:
Ask on IRC :)

asymmetric commented on components/pkg-export-helm/src/chart.rs:
```diff
@@ -125,7 +125,7 @@ impl<'a> Chart<'a> {
     }
 
-    pub fn generate(&mut self) -> Result<()> {
+    pub fn generate(mut self) -> Result<()> {
```
I wonder if this should be named differently, given how it consumes `self`. Maybe something like `convert`?

zeenix replied:
I don't think so. Maybe when `TryInto` trait is in stable `std` crate, we should instead implement `TryInto<String>` here.

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -80,9 +80,20 @@ impl Manifest {
 
+        let tag = match pkg_ident.version {
+            Some(v) => {
+                match pkg_ident.release {
```
I wonder if there's a way to do this with a chain of functions, something like: `pkg_ident.release.if_present(|r| format!(...))`.

Basically a way to inline the `match`.

zeenix replied:
for transforming the content of `Some`, there is Option::map() so i think it's doable. I'll check it out.

zeenix replied:
i managed to get rid of the inner match (against release) but couldn't manage to do the same for the outer one w/o involving a string copy.

asymmetric reviewed (changes requested):
Mostly non-blocking comments.

Commit:
pkg-export-docker: Drop noop 'latest' tag options

Docker always creates the 'latest' tag so the option to disable 'latest'
tag is completely noop. Let's remove this option.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-docker: export_for_cli_matches() returns DockerImage

If the generated image wasn't deleted, this convenience function now
returns the DockerImage instance that contains info about the image.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Get docker image info from docker exporter

Instead of assuming the name of the docker image, let's get that
information from the docker exporter itself.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

zeenix commented:
Force push done!

asymmetric commented on components/pkg-export-docker/src/main.rs:
```diff
@@ -43,5 +43,5 @@ fn start(ui: &mut UI) -> Result<()> {
     debug!("clap cli args: {:?}", m);
 
-    export_docker::export_for_cli_matches(ui, &m)
+    export_docker::export_for_cli_matches(ui, &m).map(|_| ())
```
It seems like nothing in the `start` function raises an error, so does its return type make sense? Maybe it should just have none, and then you woudn't have to use this `map` - which looks kinda weird, you're basically saying: iterate over every result, and completely ignore them.

krnowak replied:
`export_to_cli_matches` used to return `Result<()>`, now it returns `Result<DockerImage>`, so we need to turn it back to `Result<()>`. This can be done that way or with `.and(Ok(()))`, or with:
```
export_docker::export_for_cli_matches(ui, &m)?;
Ok(())
```

zeenix replied:
@asymmetric what @krnowak said and 'map' on Result and Option doesn't iterate, just maps one content to another.

I like the `and(Ok())` better than the explicit separate return.

zeenix replied:
After changing, I'm no longer sure it looks better:

-    export_docker::export_for_cli_matches(ui, &m).map(|_| ())
+    export_docker::export_for_cli_matches(ui, &m).and(Ok(()))

meh. :)

asymmetric replied:
I think the second one conveys the meaning better.

zeenix replied:
hmm... maybe but the ugliness of `Ok(())` annoys me more. :)

asymmetric replied:
It's beautiful.

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -91,9 +95,26 @@ impl Manifest {
+                    Some(i) => {
+                        (
+                            i.name().to_owned(),
+                            i.tags().get(0).unwrap_or(&"latest".to_owned()).clone(),
```
Out of curiosity, why do you need the `&`? Isn't a `&str` already a borrow?

Also, why do you need the `clone`? Rust is so hard sometimes 😭

krnowak replied:
`tags()` returns `&[String]` (reference to the slice of Strings). `get` on the slice returns something like `Option<&String>`.

`unwrap_or` for `Option<T>` looks like `fn unwrap_or(self, def: T) -> T`, so in our case where `T` is `&String`, `unwrap_or` is `fn unwrap_or(self, def: &String) -> &String`. We need to pass a reference to String to `unwrap_or`.

`&` in `&"latest".to_owned()` borrows a reference not to `"latest"` but to the result of `"latest".to_owned()`. `to_owned()` clones and returns `String` so that's why prepend `&` there, because we need `&String` here.

`unwrap_or` returns `&String`, so this stuff probably causes lifetime issues, so we need to clone it to get `String`. So if there was not tag at all we are cloning the cloned string. Whoop dee doo.

krnowak replied:
The double clone probably could be avoided with:
```
match i.tags().get(0) {
    Some(ref s) => s.clone(),
    None => "latest".to_owned(),
}
```

zeenix replied:
Thanks for explaining @krnowak . But there is no actual double clone here. `&"latest.to_owned()` is only happening if value is None, so your match statement actually does exactly the same.

asymmetric replied:
Asked on IRC, this was presented as a better solution:

`i.tags().get(0).cloned().unwrap_or_else(|| "latest".to_owned())`.

2 changes:

* [`cloned()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.cloned) - From `Option<&T>` to `Option<T>`. Kind of a syntactic sugar to remove the last `clone()`. 
* `unwrap_or_else` instead of `unwrap_or` - since it's the body of a function, it's only executed if needed. Previously, the `&str` was being cloned every time.

krnowak replied:
This solution sounds nice. Better than current one, because `"latest".to_owned()` is executed every time now - it is just a parameter to a function call.

asymmetric replied:
Forgot to mention that `cloned()` also allows us to remove the ugly `&` in `&"latest"` 👍

asymmetric commented on components/pkg-export-kubernetes/src/manifest.rs:
```diff
@@ -91,9 +95,26 @@ impl Manifest {
         let name = format!("{}-{}", pkg_ident.name, version_suffix);
 
-        let image = match matches.value_of("IMAGE_NAME") {
+        let image_name = match matches.value_of("IMAGE_NAME") {
```
One way to kill another `match`, which I think would work:

```
value_of("IMAGE_NAME")
.map(|i| i.to_owned())
.or_else(<None block>)
```

Not sure if it's pushing it too far though 🤷‍♂️

krnowak replied:
I wonder what's wrong with `match`…

Or are we playing perlgolf? Err, I mean - rustgolf?

zeenix replied:
nothing wrong with `match` really and it's one of the coolest thing after sliced bread but I've seen rust devs prefering `Option` & `Result` method chaining over match and usually it makes things very compact so why not. However, there is no reason to be go on a "must kill all match statements" spree. :)

asymmetric replied:
Agreed, I've also had the impression that chaining is more idiomatic - part of the charm of rust are these functional programming inspired bits 👅

krnowak committed:
Drop a file with unused code

The file was confusing as the lot of code there was very similar to
the code in lib.rs, but with some small differences.

Also, the net.rs file was not mentioned in lib.rs, means it wasn't
even built.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

thesentinels committed:
Merge pull request #4464 from habitat-sh/blog/getting-started-node

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #4448 from kinvolk/zeenix/k8s-tests

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #4481 from kinvolk/krnowak/drop-old-file

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

Commit:
pkg-export-k8s: Don't hardcode version in plan.sh

Hardcoding it apparently makes it impossible for the builder to find the
kubernetes exporter.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>
